### PR TITLE
[MC,ELF] Emit warning if a string constant contains newline char

### PR DESCRIPTION
**Pull request by chestnykh**
GAS emits warning about newline in the string constant so make the same behaviour.

### CODE REVIEW

**Review by s-barannikov — DISMISSED** [review]
1. Printing text to stdout is not the correct way to report warnings, it should be done via source manager (and it is likely there are helper functions).
2. AsmLexer is a lexer, not parser. It does not know the context the lexed string is used. In some contexts the warning may not be desired. The warning, if needed, should be emitted by a parser that knows full context, e.g. in `AsmParser::parseDirectiveAscii`.

**Comment by chestnykh** [comment]
> 1. Printing text to stdout is not the correct way to report warnings, it should be done via source manager (and it is likely there are helper functions).
> 
>     2. AsmLexer is a lexer, not parser. It does not know the context the lexed string is used. In some contexts the warning may not be desired. The warning, if needed, should be emitted by a parser that knows full context, e.g. in `AsmParser::parseDirectiveAscii`.

I reworked

File: llvm/lib/MC/MCParser/AsmLexer.cpp
Comment (s-barannikov) [inline llvm/lib/MC/MCParser/AsmLexer.cpp:650]: I guess the comment can just be removed.

File: llvm/test/MC/ELF/warn-newline-in-string-constant.s
Comment (s-barannikov) [inline llvm/test/MC/ELF/warn-newline-in-string-constant.s:26]: FileCheck is by default space-insensitive. If you want to test the location of a caret, you need to add `--strict-whitespace` argument.

File: llvm/lib/MC/MCParser/AsmParser.cpp
Comment (s-barannikov) [inline llvm/lib/MC/MCParser/AsmParser.cpp:3132]: The location will be incorrect if the string contains escaped characters (e.g. '\xFF'), and I don't see a way how you can reliably get the location of a newline without making `parseEscapedString` return it. Moreover, the string could contain '\n', which must not be diagnosed.

Comment (s-barannikov) [reply llvm/lib/MC/MCParser/AsmParser.cpp:3132]: I think it may make sense to look for newlines and emit warnings inside `parseEscapedString`. If not all callers of the function want these warnings, they can be issues conditionally based on a new boolean argument, but I couldn't find such a caller.

File: llvm/lib/MC/MCParser/AsmParser.cpp
Comment (s-barannikov) [inline llvm/lib/MC/MCParser/AsmParser.cpp:3135]: Ideally, you wold want to check what `Warning` returns. It can return `true` in --fatal-warning mode, in which case compilation should fail.

Comment (chestnykh) [reply llvm/lib/MC/MCParser/AsmParser.cpp]: Done

File: llvm/test/MC/ELF/warn-newline-in-string-constant.s
Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-string-constant.s:26]: Done

File: llvm/lib/MC/MCParser/AsmParser.cpp
Comment (chestnykh) [reply llvm/lib/MC/MCParser/AsmParser.cpp]: Moved the code to `parseEscapedString`. Seems that all cases are handled correctly, i've tried to write tests to each of them.

File: llvm/lib/MC/MCParser/AsmParser.cpp
Comment (s-barannikov) [inline llvm/lib/MC/MCParser/AsmParser.cpp:3132]: The correct style is:
```suggestion
      if (parseEscapedString(Data, /*WarnNewline=*/true))
```

File: llvm/lib/MC/MCParser/DarwinAsmParser.cpp
Comment (s-barannikov) [inline llvm/lib/MC/MCParser/DarwinAsmParser.cpp:616]: I guess the intent of the warning is to help finding missing quotes in such pieces of assembly:

```asm
.string "asdf
// lots of directives / code here
.string "qwer"
```

The assembler would complain about unknown instruction mnemonic "qwer" because it parsed everything after the very first quote as an argument of the first `.string` directive. This consideration should apply to `.linker_option` directive as well, which is unlikely to expect newlines in the argument. That is, I would prohibit newlines in the argument of most, if not all, directives calling `parseEscapedString`.

File: llvm/test/MC/ELF/warn-newline-in-string-constant.s
Comment (s-barannikov) [inline llvm/test/MC/ELF/warn-newline-in-string-constant.s:64]: Missing newline at end of file.

File: llvm/lib/MC/MCParser/AsmParser.cpp
Comment (s-barannikov) [inline llvm/lib/MC/MCParser/AsmParser.cpp:3034]: I think you can reduce this to just the following, but need to check:
```suggestion
        SMLoc NewlineLoc = SMLoc::getFromPointer(Str.data() + i);
```

File: llvm/test/MC/ELF/warn-newline-in-string-constant.s
Comment (s-barannikov) [inline llvm/test/MC/ELF/warn-newline-in-string-constant.s:1]: Discarded output should go to /dev/null

File: llvm/lib/MC/MCParser/DarwinAsmParser.cpp
Comment (chestnykh) [reply llvm/lib/MC/MCParser/DarwinAsmParser.cpp:616]: So your suggestion is to emit `Error()` instead of `Warning()` for all such directives ?

Comment (chestnykh) [reply llvm/lib/MC/MCParser/DarwinAsmParser.cpp:616]: I agree with you. Such a code with newlines in strange places most likely can only appear if it is written manually.
So i i think that prohibiting of newlines in all `parseEscapedString()` places is a good idea

Comment (chestnykh) [reply llvm/lib/MC/MCParser/DarwinAsmParser.cpp:616]: And if it is just a warning, it will most likely be ignored

Comment (chestnykh) [reply llvm/lib/MC/MCParser/DarwinAsmParser.cpp:616]: > No, I mean the warning should be issued for most/all of such directives, not only for .string/.ascii. Sorry if I wasn't clear.

No problem :)
my understanding was very close.
Okay, then i'm going to remove the second argument of `parseEscapedString` and emit warning unconditionally.

Comment (chestnykh) [reply llvm/lib/MC/MCParser/DarwinAsmParser.cpp:616]: Then i should extend test cases

Comment (s-barannikov) [reply llvm/lib/MC/MCParser/DarwinAsmParser.cpp:616]: A couple of different directives should be enough, other directives do not add new context.

File: llvm/test/MC/ELF/warn-newline-in-string-constant.s
Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-string-constant.s]: Done

File: llvm/lib/MC/MCParser/AsmParser.cpp
Comment (chestnykh) [reply llvm/lib/MC/MCParser/AsmParser.cpp]: Yes, it works, thanks

File: llvm/test/MC/ELF/warn-newline-in-string-constant.s
Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-string-constant.s]: Done

File: llvm/lib/MC/MCParser/DarwinAsmParser.cpp
Comment (chestnykh) [reply llvm/lib/MC/MCParser/DarwinAsmParser.cpp]: Done

File: llvm/lib/MC/MCParser/AsmParser.cpp
Comment (MaskRay) [inline llvm/lib/MC/MCParser/AsmParser.cpp:3133]: revert unneeded change

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (MaskRay) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:21]: Instead of string/string/string check/check/check, consider check/string check/string check/string that new tests use more.

You can use the following to test relative line numbers, so that inserting lines in the future would not require test update. We often omit the filename.
```
[[#@LINE+1]]:21: warning: 
```

File: llvm/lib/MC/MCParser/AsmParser.cpp
Comment (chestnykh) [reply llvm/lib/MC/MCParser/AsmParser.cpp]: Done

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s]: Thx, reworked

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (MaskRay) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:42]: You can delete `^` checks. With column numbers on the above line, `^` check isn't useful.

Use `-NEXT` whenever applicable

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (MaskRay) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:63]: You can use `FileCheck --implicit-check-not=warning:` and remove these `-NOT:` patterns

Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s]: Done

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s]: Done

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (MaskRay) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:1]: Just use x86_64 as the triple to emphasize that the behavior is for all ELF (actually other object file formats as well, but we don't bother testing), not just Linux.

**Review by MaskRay — APPROVED** [review]

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s]: Done. I don't have write access so please merge this PR when it is ready

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (s-barannikov) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:1]: There should be just one --implicit-check-not with a regex for the undesired warning. This will work as long as all expected warnings are checked by CHECK lines.

Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s:1]: like `--implicit-check-not=valid*_string` ?

Comment (s-barannikov) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s:1]: E.g. `--implicit-check-not="{{[0-9]+:[0-9]+: warning: unterminated string}}"`

Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s:1]: Thanks, got it

Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s]: Done

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (s-barannikov) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:17]: These CHECK-NEXT directives check nothing because you explicitly renamed the default "CHECK" to "CHECK-WARN" by --check-prefix option. The correct directive would therefore be "CHECK-WARN-NEXT". 

--check-prefix is only useful when you have multiple RUN lines. Using it when you have single RUN line does not hurt, but is just redundant.

(I'm not sure it is worth checking what these -NEXT directives are supposed to check. I'd just remove them, but you can keep them if you like, just fix the spelling of the directives.)

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (s-barannikov) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:1]:
```suggestion
// RUN: llvm-mc -filetype=obj -triple x86_64 %s 2>&1 -o /dev/null \
//   | FileCheck %s --implicit-check-not="{{[0-9]+:[0-9]+: warning: unterminated string}}" --check-prefix=CHECK-WARN
```

-DFILE is unused and --strict-whitespace no longer seems necessary (?)
Long RUN lines can be split into multiple by the backslash character.

Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s]: Done

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s:18]: Done

**Review by s-barannikov — APPROVED** [review]
LGTM

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (MaskRay) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:2]: We usually just test `--implicit-check-not=warning:`

**Comment by s-barannikov** [comment]
I've rebased the branch to check if that would resolve CI failure (it looked unrelated).

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s]: Done

File: llvm/test/MC/ELF/warn-newline-in-escaped-string.s
Comment (MaskRay) [inline llvm/test/MC/ELF/warn-newline-in-escaped-string.s:2]: continuation lines should be indented by 2

Comment (chestnykh) [reply llvm/test/MC/ELF/warn-newline-in-escaped-string.s]: Done

**Pull request merged**
